core/src/main/scala/kafka/log/Log.scala

```diff
@@ -74,9 +84,33 @@ case class LogAppendInfo(var firstOffset: Long,
                          validBytes: Int,
                          offsetsMonotonic: Boolean,
                          producerAppendInfos: Map[Long, ProducerAppendInfo],
+                         completedTransactions: List[CompletedTxn] = null,
                          isDuplicate: Boolean = false)
 
 /**
+ * A class used to hold useful metadata about a completed transaction. This is used to build
+ * the transaction index after appending to the log.
+ *
+ * @param producerId The ID of the producer
+ * @param firstOffset The first offset (inclusive) of the transaction
+ * @param lastOffset The last offset (inclusive) of the transaction. This is always the offset of the
+ *                   COMMIT/ABORT control record which indicates the transaction's completion.
+ * @param isAborted Whether or not the transaction was aborted
+ */
+case class CompletedTxn(producerId: Long,
+                        firstOffset: Long,
+                        lastOffset: Long,
+                        isAborted: Boolean) extends Ordered[CompletedTxn] {
+  override def compare(that: CompletedTxn): Int = {
+    val res = this.firstOffset compare that.firstOffset
```

Comment:
Sorting by the first offset is what we want. This is used to find the first unstable offset which will always be the minimum first offset of all transactions. I couldn't find the comment you were referring to since the patch was updated. If it still exists, can you point me to it?